code/parachain/frame/cosmwasm/src/tests.rs

```diff
@@ -1,6 +1,221 @@
+
+		assert!(Cosmwasm::do_ed25519_batch_verify(
+			&ref_messages,
+			&ref_signatures,
+			&ref_public_keys
+		));
+	})
+}
+
+#[test]
+fn ed25519_batch_verify_verifies_multisig() {
+	new_test_ext().execute_with(|| {
+		let decode = |m| -> Vec<u8> { hex::decode(m).unwrap() };
+
+		let messages: Vec<Vec<u8>> = [ED25519_MESSAGE_HEX].iter().map(decode).collect();
+		let signatures: Vec<Vec<u8>> =
+			[ED25519_SIGNATURE_HEX, ED25519_SIGNATURE_HEX].iter().map(decode).collect();
+		let public_keys: Vec<Vec<u8>> =
+			[ED25519_PUBLIC_KEY_HEX, ED25519_PUBLIC_KEY_HEX].iter().map(decode).collect();
+
+		let ref_messages: Vec<&[u8]> = messages.iter().map(|b| b.as_slice()).collect();
+		let ref_signatures: Vec<&[u8]> = signatures.iter().map(|b| b.as_slice()).collect();
+		let ref_public_keys: Vec<&[u8]> = public_keys.iter().map(|b| b.as_slice()).collect();
+
+		assert!(Cosmwasm::do_ed25519_batch_verify(
+			&ref_messages,
+			&ref_signatures,
+			&ref_public_keys
+		));
+	})
+}
+
+#[test]
+fn ed25519_batch_verify_verifies_with_single_pubkey_multi_msg() {
+	new_test_ext().execute_with(|| {
+		let decode = |m| -> Vec<u8> { hex::decode(m).unwrap() };
+
+		let messages: Vec<Vec<u8>> =
+			[ED25519_MESSAGE_HEX, ED25519_MESSAGE_HEX].iter().map(decode).collect();
+		let signatures: Vec<Vec<u8>> =
+			[ED25519_SIGNATURE_HEX, ED25519_SIGNATURE_HEX].iter().map(decode).collect();
+		let public_keys: Vec<Vec<u8>> = [ED25519_PUBLIC_KEY_HEX].iter().map(decode).collect();
+
+		let ref_messages: Vec<&[u8]> = messages.iter().map(|b| b.as_slice()).collect();
+		let ref_signatures: Vec<&[u8]> = signatures.iter().map(|b| b.as_slice()).collect();
+		let ref_public_keys: Vec<&[u8]> = public_keys.iter().map(|b| b.as_slice()).collect();
+
+		assert!(Cosmwasm::do_ed25519_batch_verify(
+			&ref_messages,
+			&ref_signatures,
+			&ref_public_keys
+		));
+	})
+}
+
+#[test]
+fn ed25519_batch_verify_fails_if_one_fail() {
+	new_test_ext().execute_with(|| {
+		let decode = |m| -> Vec<u8> { hex::decode(m).unwrap() };
+
+		let messages: Vec<Vec<u8>> =
+			[ED25519_MESSAGE_HEX, ED25519_MESSAGE2_HEX].iter().map(decode).collect();
+		let mut signatures: Vec<Vec<u8>> =
+			[ED25519_SIGNATURE_HEX, ED25519_SIGNATURE2_HEX].iter().map(decode).collect();
+		let public_keys: Vec<Vec<u8>> =
+			[ED25519_PUBLIC_KEY_HEX, ED25519_PUBLIC_KEY2_HEX].iter().map(decode).collect();
+
+		*signatures.last_mut().unwrap().last_mut().unwrap() += 1;
+
+		let ref_messages: Vec<&[u8]> = messages.iter().map(|b| b.as_slice()).collect();
+		let ref_signatures: Vec<&[u8]> = signatures.iter().map(|b| b.as_slice()).collect();
+		let ref_public_keys: Vec<&[u8]> = public_keys.iter().map(|b| b.as_slice()).collect();
+
+		assert!(!Cosmwasm::do_ed25519_batch_verify(
+			&ref_messages,
+			&ref_signatures,
+			&ref_public_keys
+		));
+	})
+}
+
+#[test]
+fn ed25519_batch_verify_fails_if_input_lengths_are_incorrect() {
+	new_test_ext().execute_with(|| {
```

Comment:
The functions can be extracted from the pallet for the mock context not to be required. But it is not a big deal.